middlewares/retry_test.go

```diff
@@ -0,0 +1,92 @@
+package middlewares
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+)
+
+func TestRetry(t *testing.T) {
+	testCases := []struct {
+		failAtCalls    []int
+		retryCount     int
+		responseStatus int
+	}{
+		{
+			failAtCalls:    []int{1, 2},
+			retryCount:     3,
+			responseStatus: http.StatusOK,
+		},
+		{
+			failAtCalls:    []int{1, 2},
+			retryCount:     2,
+			responseStatus: http.StatusBadGateway,
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		tcName := fmt.Sprintf("FailAtCalls(%v) RetryCount(%v)", tc.failAtCalls, tc.retryCount)
+
+		t.Run(tcName, func(t *testing.T) {
+			t.Parallel()
+
+			var httpHandler http.Handler
+			httpHandler = &networkFailingHTTPHandler{failAtCalls: tc.failAtCalls}
+			httpHandler = NewRetry(tc.retryCount, httpHandler)
+
+			recorder := httptest.NewRecorder()
+			req := httptest.NewRequest(http.MethodGet, "http://localhost:3000/ok", ioutil.NopCloser(nil))
+
+			httpHandler.ServeHTTP(recorder, req)
+
+			if tc.responseStatus != recorder.Code {
+				t.Errorf("got status code %v, wanted %v", recorder.Code, tc.responseStatus)
+			}
+		})
+	}
+}
+
+func TestRecordNetError(t *testing.T) {
+	boolNetErrorOccurred := false
```

Comment:
A small hint for reviewers, I tried to build this as a table driven test at first, but I was not able to get it working in a straightforward way, as it would require me to specify _one_ type for the initial context value I am passing in. This would have to be an `interface{}` in my case, but then the type assertion on a `*bool` in the `DefaultRecordNetError` does not work anymore.